xarray/tests/test_combine.py

```diff
@@ -818,6 +818,38 @@ def test_combine_by_coords_incomplete_hypercube(self):
         with pytest.raises(ValueError):
             combine_by_coords([x1, x2, x3], fill_value=None)
 
+    def test_combine_by_coords_with_tolerance(self):
+        x = [0, 1, 2]
+        tol = 1e-7
+
+        x1 = x + tol * np.random.rand(3)
+        ds1 = Dataset(
+            {
+                "a": (
+                    ("time", "x"),
+                    [
+                        [
+                            9,
+                            0,
+                            2,
+                        ]
+                    ],
+                )
+            },
+            coords={"x": x1, "time": [0]},
+        )
+
+        x2 = x + tol * np.random.rand(3)
+        ds2 = Dataset(
+            {"a": (("time", "x"), [[6, 8, 3]])}, coords={"x": x2, "time": [1]}
+        )
+
+        # fail if tolerance is not properly implemented
+        combine_by_coords([ds1, ds2], tolerance=1e-6)
+
+        # fail if tolerance is not properly implemented
+        combine_by_coords([ds1, ds2], tolerance={"x": 1e-6})
```

Comment:
Good point, and in fact my PR does not work. It actually detect that x is not to be concatenated, but then combine_by_coords calls align (indirectly through a chain of function call) to deal with the other dims. Align also tests the equality between the index of the same name and use join to decide what to do. It results that the index that are close, are joined. This is not the expected behavior.

To solve this issue, the first option I see is to add a tolerance argument to align (and to concat which calls align). This is profound changes, I need to opinion on that.

The alternative option is to copy the datasets in  _infer_concat_order_from_coords and for the dims that are found to be allclose in all the datasets, set the same index in all the datasets (i.e. taking the index of the first dataset, and set it to all the others). This will trick the next step when concat and align are called and compares the indexes. I don't really like this option but it has the advantage to minimize the changes, and avoid touching concat and align. Adding a tolerance to concat and align is certainly useful, as it is for combine_by_coords, but it must be done well...

What do you think ?

Reply:
I think both of these options would be OK.

With regards to the first solution (adding tolerance to align), that's exactly what we agreed on in https://github.com/pydata/xarray/issues/2217. But indeed, that would be a larger change (probably best handled separately).

With regards to the second solution, I suspect you could call `align` with `join='override'` in the right place to set the indexes properly. The challenge would be tracking down where that needs to be set, and ensuring that it doesn't effect other unrelated code paths.